### DISCUSSION TITLE
better-sqlite3 works under node20 but gives error under deno

### DISCUSSION THREAD

**Discussion by bkerin** [denoland/deno#24258] [General]
I'm not sure it's a bug but with minimal node20 test project this `index.mjs` run with `node index.mjs` after `node install better-sqlite3` works as expected:
```
port sqlite from 'better-sqlite3'
const db = sqlite()
console.log(db)
```
while this near-equivalent `test_better_sqlite3.js` file run with `deno run -A test_better_sqlite3.js` fails (errors below):
```
import sqlite from "npm:better-sqlite3"
const db = sqlite()
console.log(db)
```

The error output is:
```
error: Uncaught (in promise) Error: Could not locate the bindings file. Tried:
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/build/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/build/Debug/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/build/Release/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/out/Debug/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/Debug/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/out/Release/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/Release/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/build/default/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/compiled/20.11.1/linux/x64/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/addon-build/release/install-root/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/addon-build/debug/install-root/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/addon-build/default/install-root/better_sqlite3.node
 → /home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/lib/binding/node-v108-linux-x64/better_sqlite3.node
    at bindings (file:///home/bkerin/.cache/deno/npm/registry.npmjs.org/bindings/1.5.0/bindings.js:126:9)
    at new Database (file:///home/bkerin/.cache/deno/npm/registry.npmjs.org/better-sqlite3/11.0.0/lib/database.js:48:64)
    at Database (file:///home/bkerin/.cache/deno/npm/registry.npmjs.org/better-sqlite3/11.0.0/lib/database.js:11:10)
    at file:///home/bkerin/projects/cell_electric_meter/web_server/test_better_sqlite3.js:3:12
```

Is this expected or might it somehow be due to use of node module or npm specifier or something?

**Comment by bkerin** [comment]
It looks like it's a deno bug in npm compat layer https://github.com/denoland/deno/issues/15717

**Comment by bkerin** [comment]
Poking around in lucia-auth+adapter-sqlite it looks like `jsr:@db/sqlite@0.11.1` has an identical API at least as far as `@lucia-auth+adapter-sqlite@3.0.1/node_modules/@lucia-auth/adapter-sqlite/dist/drivers` goes.  So adding support for it to `lucia-auth+sqlite` is probably really trivial and might be a good idea if fixing the problem in the npm compat layer is too hard.  I haven't figured out how module publication/fork/override work in deno/node/javascript land yet myself but for someone who knows it I think it would be super easy.

Then maybe there would be a reasonably working auth setup for my situation (deno+Hono behind a reverse proxy).  So far auth.js fails because Hono can't handle the reverse proxy and and Lucia fails as above.

**Comment by alexgleason** [comment]
You might want to use [Kysely](https://kysely.dev/) so you can write your code one way and switch out the adapter if needed.

I created a [kysely-deno-sqlite](https://jsr.io/@soapbox/kysely-deno-sqlite) that works with both @db/sqlite and x/sqlite

**Comment by bkerin** [comment]
TBH I'm sort of struggling to stuff all the changes that have happened in web devel over the last 20 years into my head at once and though kysely sounds like a good idea I'd rather use the sql I know if possible.

This page: https://lucia-auth.com/database/kysely
shows BetterSqlite3Adapter still being used with Kysely I guess your wrapper could be used in place pretty easily I'd probably struggly a bit to figure out exactly how.

**Comment by Mutefish0** [comment]
https://github.com/denoland/deno/issues/24740
